### PR TITLE
Gui

### PR DESCRIPTION
**Pull request by Qozul**
Gui has slightly unstable, but working and reasonably consistent descriptor management for images, push constants, and buffers. Refactoring and further information will be added soon, but this version should provide a reasonable addition to master.

### CODE REVIEW

File: Vulkan-Pipeline-Assistant/mainwindow.h [hunk 1]
```diff
@@ -55,14 +57,16 @@ namespace vpa {
         QWidget* MakeMultisampleBlock();
         QWidget* MakeDepthStencilBlock();
         QWidget* MakeRenderPassBlock();
+        void MakeDescriptorBlock();
 
-        QComboBox* MakeComboBox(QWidget* parent, QVector<QString> items);
+        void VulkanCreationCallback();
 
         Ui::MainWindow* m_ui;
         VulkanMain* m_vulkan;
```
Comment (Kney-Delach) [inline Vulkan-Pipeline-Assistant/mainwindow.h:65]: Add a note to trello to explicitly assign all empty data values in constructor.

File: Vulkan-Pipeline-Assistant/spvstructwidget.h [hunk 2]
```diff
@@ -0,0 +1,31 @@
+#ifndef SPVSTRUCTWIDGET_H
+#define SPVSTRUCTWIDGET_H
+
+#include "spvwidget.h"
+
+class QPushButton;
```
Comment (Kney-Delach) [inline Vulkan-Pipeline-Assistant/spvstructwidget.h:6]: For clarity maybe add a space between forward declarations and implementations.

File: Vulkan-Pipeline-Assistant/spvvectorwidget.cpp [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    m_infoGroup = new QWidget(parent);
+    m_infoGroup->setLayout(new QHBoxLayout(m_infoGroup));
+    m_infoGroup->layout()->addWidget(new QLabel(VecStrings[(m_type->length - 1)], parent));
+    QPushButton* normBtn = new QPushButton("Normalize", m_infoGroup);
+    m_infoGroup->layout()->addWidget(normBtn);
+    connect(normBtn, SIGNAL(pressed()), this, SLOT(HandleNormalize()));
+    m_infoGroup->layout()->setAlignment(Qt::AlignTop);
+
+    m_inputsGroup = new QWidget(parent);
+    QGridLayout* inputsGroupLayout = new QGridLayout(m_inputsGroup);
```
Comment (Kney-Delach) [inline Vulkan-Pipeline-Assistant/spvvectorwidget.cpp:31]: Does this memory need to be freed locally?

**Review by Kney-Delach — APPROVED** [review]
All looks good, address comment and i'm okay for you to merge this.

File: Vulkan-Pipeline-Assistant/mainwindow.h (hunk 1 above)
Comment (Qozul) [reply Vulkan-Pipeline-Assistant/mainwindow.h:65]: Complete

File: Vulkan-Pipeline-Assistant/spvvectorwidget.cpp (hunk 3 above)
Comment (Qozul) [reply Vulkan-Pipeline-Assistant/spvvectorwidget.cpp:31]: No it doesn't, Qt adds it to its own memory management system as I understand it. When the parent is destroyed, all children will be. Ultimately, the window is destroyed and it is at the top of the parent chain.

File: Vulkan-Pipeline-Assistant/spvstructwidget.h (hunk 2 above)
Comment (Qozul) [reply Vulkan-Pipeline-Assistant/spvstructwidget.h:6]: I'll add a note in trello for refactoring to check over style and ensure everything is up to par, including a note of this in its comments.

**Comment by Qozul** [comment]
Merging and closing

**Pull request merged**